### ISSUE TITLE
Intellij 2017 uses IDE bundled JRE for deployment

### ISSUE THREAD

**Issue by etanshaul** [GoogleCloudPlatform/google-cloud-intellij#1316]
For some reason, its not using the project configured JDK. This causes staging errors:

"Unable to stage app: Cannot get the System Java Compiler. Please use a JDK, not a JRE."

**Comment by patflynn** [comment]
I just successfully did an app engine standard deployment on os x with 2017 so this might not affect all users

**Comment by etanshaul** [comment]
It may have something to do with how java is configured on my system. Currently on my machine I can deploy from 2016, but not from 2017. Its good to know that this doesn't affect all users. I will still try to figure out whats causing this.

**Comment by etanshaul** [comment]
@patflynn also, when you did that test, did you pick up the new 2017 version of the plugin we just released?

**Comment by patflynn** [comment]
I think so.

**Comment by etanshaul** [comment]
were you running standalone 2017 or built via our project? what @joaoandremartins and I both noticed is that we only encounter this problem on standalone 2017 instances.

**Comment by etanshaul** [comment]
I tried standalone 2017 on my mac and it worked

**Comment by joaoandremartins** [comment]
I noticed that the 2016.3 and 2017.1 idea.sh use slightly different ways of getting the JDK.

I don't fully understand what's going on, but might have something to do with this.

First is 2017.1, then 2016.3.

71c71
< elif [ "$OS_TYPE" = "Linux" ] && [ -x "$IDE_HOME/jre/bin/java" ] && "$IDE_HOME/jre/bin/java" -version > /dev/null 2>&1 ; then
---
> elif [ -x "$IDE_HOME/jre/jre/bin/java" ] && "$IDE_HOME/jre/jre/bin/java" -version > /dev/null 2>&1 ; then

**Comment by joaoandremartins** [comment]
Could it be that IJ's latest JDK version doesn't contain "com.sun.tools.javac.api.JavacTool"?

JDK of 2017.1 where deployment fails (112-release-736-b13 amd64):

IntelliJ IDEA 2017.1
Build #IU-171.3780.107, built on March 22, 2017
Licensed to Google

JRE: 1.8.0_112-release-736-b13 amd64
JVM: OpenJDK 64-Bit Server VM by JetBrains s.r.o
Linux 3.13.0-107-generic

JDK of 2016.3 where deploy succeeds (112-release-408-b2 amd64):

IntelliJ IDEA 2016.3.1
Build #IU-163.9166.29, built on December 9, 2016
Licensed to Google

JRE: 1.8.0_112-release-408-b2 amd64
JVM: OpenJDK 64-Bit Server VM by JetBrains s.r.o

JDK of 2017.1 spun up by ./gradlew runIdea (102-b14 amd64):

IntelliJ IDEA 2017.1
Build #IU-171.3780.107, built on March 22, 2017
Licensed to Google

JRE: 1.8.0_102-b14 amd64
JVM: Java HotSpot(TM) 64-Bit Server VM by Oracle Corporation
Linux 3.13.0-107-generic

**Comment by joaoandremartins** [comment]
Opened https://youtrack.jetbrains.com/issue/IDEA-170316 to check with IJ.

**Comment by joaoandremartins** [comment]
@patflynn do you want to leave this open?

**Comment by patflynn** [comment]
I think we should leave it open until we've merged into master probably.

**Comment by joaoandremartins** [comment]
It is on "2017's master", but I'm OK either way.

**Comment by jonathanlal** [comment]
Still getting this god forsaken issue, every now and then when the Java on my computer automatically updates itself. It's incredibly annoying. And oddly, only have this problem with one of my projects but all the configuration settings are the same as the configs for my other projects.... wtf is going on.

**Comment by etanshaul** [comment]
@jonathanlal that is odd. Just to get some more info to see how we can help:

You say you get it every now and then after an update; are you able to then resolve the problem after the update? What steps did you take if so?

**Comment by jonathanlal** [comment]
I know this is an Intellij thread, but it's the same with Eclipse from what i gather. Last time it happened i fixed it like this (last answer):
[stackoverflow](https://stackoverflow.com/questions/19168082/eclipse-appengine-java-lang-runtimeexception-cannot-get-the-system-java-compi/45256356#45256356)



No luck fixing it this time. I have literally two dynamic java based web projects with the same configurations, the only difference is that one has .jsp files and the other doesn't. So that narrows down the problem significantly.  Something to do with including .jsp files in the project, produces this error:

```
Unable to stage app: Cannot get the System Java Compiler. Please use a JDK, not a JRE.
Please see the logs [C:\Users\JONATH~1\AppData\Local\Temp\appcfg7118762364559219351.log] for further information.
```
The project is on Java 1.7 but my system is running on 1.8. This worked fine until i allowed Java on my system to update itself.

**Comment by patflynn** [comment]
@loosebazooka this rings a bell.  Do we have a bug open somewhere WRT the stage jsp compiler not using the right jdk?

**Comment by patflynn** [comment]
@elharo 
@etanshaul 

We should verify we're properly setting java home for the staging command in Eclipse and IntelliJ.

**Comment by patflynn** [comment]
@jonathanlal is this happening for you in IntelliJ or Eclipse?  

@etanshaul discovered that we do seem to use the project jdk for staging.

**Comment by jonathanlal** [comment]
@patflynn

Eclipse.
I got it working again last night. But only after uninstalling all the different versions of java i had on my system. Then only using jdk7 everywhere. So i reckon it's an issue with jdk8.

**Comment by patflynn** [comment]
@chanseokoh Could you double-check that Eclipse is doing the right thing here?

**Comment by etanshaul** [comment]
I'll relocate this to the Eclipse tracker

**Comment by etanshaul** [comment]
https://github.com/GoogleCloudPlatform/google-cloud-eclipse/issues/2548
